### ISSUE TITLE
New comment on block [block-idp10740192]

### ISSUE THREAD

**Issue by gregtatcam** [realworldocaml/book#1237]
The compile fails:

ocamlbuild -package async,uri,yojson,cohttp.async,textwrap -use-ocamlfind search1.native
- ocamlfind ocamlc -c -g -annot -thread -package core -package async,uri,yojson,cohttp.async,textwrap -o search1.cmo search1.ml
  File "search1.ml", line 30, characters 26-42:
  Error: This expression has type server:Core.Std.String.t -> Uri.t
     but an expression was expected of type Uri.t
  Command exited with code 2.
  Compilation unsuccessful after building 2 targets (0 cached) in 00:00:01.
  -----------------
  (\* file: search.ml *)
  open Core.Std
  open Async.Std

(\* Generate a DuckDuckGo search URI from a query string *)
let query_uri ~server query =
  let base_uri =
    Uri.of_string (String.concat ["http://";server;"/?format=json"])
  in
  Uri.add_query_param base_uri ("q", [query])

(\* Extract the "Definition" or "Abstract" field from the DuckDuckGo results *)
let get_definition_from_json json =
  match Yojson.Safe.from_string json with
  | `Assoc kv_list ->
    let find key =
      begin match List.Assoc.find kv_list key with
      | None | Some (`String "") -> None
      | Some s -> Some (Yojson.Safe.to_string s)
      end
    in
    begin match find "Abstract" with
    | Some _ as x -> x
    | None -> find "Definition"
    end
  | _ -> None

(\* Execute the DuckDuckGo search *)
let get_definition word =
  Cohttp_async.Client.get (query_uri word)

> > = fun (_, body) ->
> >   Pipe.to_list body
> > | fun strings ->
> >   (word, get_definition_from_json (String.concat strings))

(\* Print out a word/definition pair *)
let print_result (word,definition) =
  printf "%s\n%s\n\n%s\n\n"
    word
    (String.init (String.length word) ~f:(fun _ -> '-'))
    (match definition with
    | None -> "No definition found"
    | Some def ->
      String.concat ~sep:"\n"
        (Wrapper.wrap (Wrapper.make 70) def))

(\* Run many searches in parallel, printing out the results after they're all
   done. *)
let search_and_print words =
  Deferred.all (List.map words ~f:get_definition)

> > | fun results ->
> >   List.iter results ~f:print_result

let () =
  Command.async_basic
    ~summary:"Retrieve definitions from duckduckgo search engine"
    Command.Spec.(
      empty
      +> anon (sequence ("word" %: string))
    )
    (fun words () -> search_and_print words)
  |> Command.run
